src/pid/maps.rs

```diff
@@ -0,0 +1,243 @@
+//! Process memory mappings information from `/proc/[pid]/maps`.
+
+use std::ffi::OsString;
+use std::io::{self, BufRead};
+use std::os::unix::ffi::OsStringExt;
+use std::path::PathBuf;
+use std::{fs, ops};
+
+use libc;
+use nom::{rest, space};
+
+use parsers::{map_result, parse_usize_hex, parse_u32_hex, parse_u64, parse_u64_hex};
+use unmangle::unmangled_path;
+
+/// Process memory mapping information.
+///
+/// Due to the way paths are encoded by the kernel before exposing them in
+/// `/proc/[pid]/maps`, the parsing of `path` and `is_deleted` is
+/// ambiguous. For example, all the following path/deleted combinations:
+///
+/// - `/tmp/a\nfile` *(deleted file)*
+/// - `/tmp/a\nfile (deleted)` *(existing file)*
+/// - `/tmp/a\\012file` *(deleted file)*
+/// - `/tmp/a\\012file (deleted)` *(existing file)*
+///
+/// will be mangled by the kernel and decoded by this module as:
+///
+/// ```rust,ignore
+/// MemoryMapping (
+///    ...,
+///    path: PathBuf::from("/tmp/a\nfile"),
+///    is_deleted:true,
+/// )
+/// ```
+///
+/// If the `path` of a mapping is required for other than purely informational
+/// uses (such as opening and/or memory mapping it), a more reliable source
+/// (such as `/proc/[pid]/map_files`) should be used, if available. The open
+/// file `(dev, inode)` should also be checked against the values provided by
+/// the mapping.
+///
+/// See `man 5 proc`, `Linux/fs/proc/task_mmu.c`, and `Linux/fs/seq_file.c`.
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct MemoryMapping {
+    /// Address range that this mapping occupies in the process virtual memory
+    /// space.
+    pub range: ops::Range<usize>,
+    /// Whether pages in this mapping may be read.
+    pub is_readable: bool,
+    /// Whether pages in this mapping may be written.
+    pub is_writable: bool,
+    /// Whether pages in this mapping may be executed.
+    pub is_executable: bool,
+    /// Whether this mapping is shared.
+    pub is_shared: bool,
+    /// Offset into the file backing this mapping (for non-anonymous mappings).
+    pub offset: u64,
+    /// Device containing the file backing this mapping (for non-anonymous
+    /// mappings).
+    pub dev: libc::dev_t,
+    /// Inode of the file backing this mapping (for non-anonymous mappings).
+    pub inode: libc::ino_t,
+    /// Path to the file backing this mapping (for non-anonymous mappings),
+    /// pseudo-path (such as `[stack]`, `[heap]`, or `[vdso]`) for some special
+    /// anonymous mappings, or empty path for other anonymous mappings.
+    pub path: PathBuf,
+    /// Whether the file backing this mapping has been deleted (for
+    /// non-anonymous mappings).
+    pub is_deleted: bool,
```

Comment:
I can't find in the linux source where `(deleted)` is added to the file path, could you point that out to me?  The manual doesn't mention it either.

Reply:
The ` (deleted)` suffix is part of the `dcache` interface. Pathnames are generated in `/proc` files using [seq_file_path()](https://grokbit.com/torvalds/linux/ref/master/fs/seq_file.c?fn=seq_file_path&hit=1), which ends up calling [d_path()](https://grokbit.com/torvalds/linux/ref/master/fs/dcache.c?q=d_path&hit=3) to generate the path for the corresponding dentry. The latter function is the one appending ` (deleted)` if the entry has been deleted.

I have added a reference to `Linux/fs/dcache.c` in the `MemoryMap` documentation. Thanks for pointing this out! (It was indeed not obvious).

Reply:
Sounds good, thanks.